### PR TITLE
[main][feat] EcoPythCalldataBuilder3, CIX support

### PR DESCRIPTION
**Pull request by kim-hera**
## Desc
Implement a EcoPythCalldataBuilder3. Supporting CIX. 
(CIX = Customized Index)

image

### CODE REVIEW

File: src/oracles/CIXPythAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        pyth.getPriceUnsafe(_pythPriceIds[i]);
+
+        unchecked {
+          ++i;
+        }
+      }
+
+      if (_weightSum != 1e8) revert CIXPythAdapter_BadWeightSum();
+    }
+
```
Comment (shutsuwei) [inline src/oracles/CIXPythAdapter.sol:203]: Should have a variation check as C is a critical variable which can change the index prices.

File: src/oracles/CIXPythAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import { ABDKMath64x64 } from "@hmx/libraries/ABDKMath64x64.sol";
+import { HMXLib } from "@hmx/libraries/HMXLib.sol";
+import { OwnableUpgradeable } from "@openzeppelin-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import { PythStructs } from "pyth-sdk-solidity/IPyth.sol";
+import { ICIXPythAdapter } from "./interfaces/ICIXPythAdapter.sol";
+import { IReadablePyth } from "./interfaces/IReadablePyth.sol";
+
+/// @dev Customized Index Pyth Adapter - Index price will be calculated using geometric mean according to weight config
+contract CIXPythAdapter is OwnableUpgradeable, ICIXPythAdapter {
```
Comment (shutsuwei) [inline src/oracles/CIXPythAdapter.sol:15]: I think this design might not be very gas effective as every read needs to calculate the index price.

I'm leaning towards make this as a EcoPythCalldataBuilder and feed their instead.

Comment (kim-hera) [reply src/oracles/CIXPythAdapter.sol:15]: Will turn this into `CIXPriceLens`, and plug it beside `OnChainPriceLens` at `EcoPythCalldataBuilder3`.

File: src/oracles/CIXPriceAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  // errors
+  error CIXPriceAdapter_MissingPriceFromBuildData();
+  error CIXPriceAdapter_BadParams();
+  error CIXPriceAdapter_BadWeightSum();
+  error CIXPriceAdapter_COverDiff();
+
+  // state variables
+  ICIXPriceAdapter.CIXConfig public config;
+  uint32 public maxCDiffBps = 1000; // 10%
```
Comment (john-xina-p88) [inline src/oracles/CIXPriceAdapter.sol:27]: immutable for gas saving

Comment (kim-hera) [reply src/oracles/CIXPriceAdapter.sol:27]: Cannot, as it is mutable via setter.

**Review by john-xina-p88 — APPROVED** [review]

**Review by shutsuwei — APPROVED** [review]

**Pull request merged**